cli/src/cmd/forge/coverage.rs

```diff
@@ -0,0 +1,283 @@
+                    }
+
+                    map.add_source(path.clone(), versioned_source, items);
+                }
+            }
+        }
+
+        Ok((map, source_maps))
+    }
+
+    /// Runs tests, collects coverage data and generates the final report.
+    fn collect(
+        self,
+        project: Project,
+        output: ProjectCompileOutput,
+        source_maps: SourceMaps,
+        mut map: CoverageMap,
+        config: Config,
+        evm_opts: EvmOpts,
+    ) -> eyre::Result<()> {
+        // Setup the fuzzer
+        // TODO: Add CLI Options to modify the persistence
+        let cfg = proptest::test_runner::Config {
+            failure_persistence: None,
+            cases: config.fuzz_runs,
+            max_local_rejects: config.fuzz_max_local_rejects,
+            max_global_rejects: config.fuzz_max_global_rejects,
+            ..Default::default()
+        };
+        let fuzzer = proptest::test_runner::TestRunner::new(cfg);
+        let root = project.paths.root;
+
+        // Build the contract runner
+        let evm_spec = crate::utils::evm_spec(&config.evm_version);
+        let mut runner = MultiContractRunnerBuilder::default()
+            .fuzzer(fuzzer)
+            .initial_balance(evm_opts.initial_balance)
+            .evm_spec(evm_spec)
+            .sender(evm_opts.sender)
+            .with_fork(utils::get_fork(&evm_opts, &config.rpc_storage_caching))
+            .with_coverage()
+            .build(root.clone(), output, evm_opts)?;
+        let (tx, rx) = channel::<(String, SuiteResult)>();
+
+        // Set up identifier
+        let local_identifier = LocalTraceIdentifier::new(&runner.known_contracts);
+
+        // TODO: Coverage for fuzz tests
+        let handle = thread::spawn(move || runner.test(&self.filter, Some(tx), false).unwrap());
+        for mut result in rx.into_iter().flat_map(|(_, suite)| suite.test_results.into_values()) {
+            if let Some(hit_map) = result.coverage.take() {
+                for (_, trace) in &mut result.traces {
+                    local_identifier
+                        .identify_addresses(trace.addresses().into_iter().collect())
+                        .into_iter()
+                        .filter_map(|identity| {
+                            let artifact_id = identity.artifact_id?;
+                            let source_map = source_maps.get(&artifact_id)?;
+
+                            Some((artifact_id, source_map, hit_map.get(&identity.address)?))
+                        })
+                        .for_each(|(id, source_map, hits)| {
+                            // TODO: Distinguish between creation/runtime in a smart way
+                            map.add_hit_map(id.version.clone(), &source_map.0, hits.clone());
+                            map.add_hit_map(id.version, &source_map.1, hits.clone())
+                        });
+                }
+            }
+        }
+
+        // Reattach the thread
+        let _ = handle.join();
```

Comment:
why execute this on a new thread and immediately join it again?

Reply:
This is how forge test works too - after some changes where we stream in results of test runs we have to run it in a new thread, otherwise we get locked I think?